win32ss/gdi/ntgdi/freetype.c

```diff
@@ -2291,9 +2289,104 @@ IntGdiAddFontResourceEx(PUNICODE_STRING FileName, DWORD Characteristics,
 }
 
 INT FASTCALL
-IntGdiAddFontResource(PUNICODE_STRING FileName, DWORD Characteristics)
+IntGdiAddFontResourceEx(
+    IN const UNICODE_STRING *FileName,
+    IN DWORD Characteristics,
+    IN DWORD dwFlags,
+    IN DWORD cFiles,
+    IN DWORD cwc)
 {
-    return IntGdiAddFontResourceEx(FileName, Characteristics, 0);
+    PWSTR pchFile = FileName->Buffer;
+    SIZE_T cchFile;
+    INT ret = 0;
+
+    while (cFiles--)
+    {
+        _SEH2_TRY
+        {
+            // Security issue: FileName should be terminated by double '\0'
+            if (!*pchFile)
+                return FALSE;
+
+            cchFile = wcslen(pchFile);
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+        {
+            _SEH2_YIELD(return FALSE);
+        }
+        _SEH2_END;
+
+        // Security issue: Prohibit buffer overrun
+        if (pchFile + (cchFile + 1) > FileName->Buffer + cwc)
+            return FALSE;
+
+        UNICODE_STRING ustrPathName;
+        ustrPathName.Length = (USHORT)(cchFile * sizeof(WCHAR));
+        ustrPathName.MaximumLength = ustrPathName.Length + sizeof(WCHAR);
+        ustrPathName.Buffer = pchFile;
+
+        INT count = IntGdiAddFontResourceSingle(&ustrPathName, Characteristics, dwFlags);
+        if (!count)
+            return 0;
+        ret += count;
+
+        pchFile += cchFile + 1;
+    }
+
+    return ret;
+}
+
+static BOOL FASTCALL
+IntGdiRemoveFontResourceSingle(
+    IN const UNICODE_STRING *FileName,
```

Comment:
```suggestion
    IN PCUNICODE_STRING FileName,
```
and same comment as for the other function.